### ISSUE TITLE
Compare results between two Robocop runs

### ISSUE THREAD

**Issue by bhirsz** [MarketSquare/robotframework-robocop#489]
It would be good to have some way of comparing two Robocop runs ie number of new/fixed issues. It could be special report (--report compare).

Other cli tools usually create special directory (like .pylint or .coverage) and store run results there - which is accessed in 2nd and evaluated in 2nd run.

**Comment by bhirsz** [comment]
If we rely on external tools, we can use SARIF results with the sarif tool:

```
dir>sarif diff .sarif.json .sarif2.json
error level: +0 -0 no changes
warning level: +0 +0
  Number of occurrences 35 -> 34 (-1) for issue "0702 Missing blank space after comment character"
note level: +0 -0 no changes
all levels: +0 +0
```

**Comment by bhirsz** [comment]
@mnojek what do you think about always creating . dot file in the current directory same way some of the tools are doing? (like .pylint or .coverage). It would be our internal file that could be used for ``--reports compare_runs``. 

OR 

we could add ``compare_runs`` (name temporary) report that would generate the file during 1st run, and then use it for comparison during 2nd run. It would require users to configure this report for 1st run if they want it to compare it with 2nd run but I like it bit more because we don't create trash files if the user doesn't want it.

The file itself could be just json-like file (we can reuse json_report for that), and ``compare_runs`` would compare the content of the file with internal_json_report issues and produce summary. Summary could be like the above: issues by severity and +- changes, and list of rules that were removed/added from report between runs.

**Comment by mnojek** [comment]
I see the benefits of both approaches, but I agree that the first one would produce some files users would not want to. On the other hand, it's more "automatic" approach, since they don't need to do anything manually, and we can compare execution results every time without even asking the users.

Regarding the 2nd approach – what would the user need to do, to compare between the 2nd and the 3rd run? I assume that if I explicitly create a JSON report for one run, all other runs will always be compared with this one, right? So the 2nd run will be compared with 1st, and the 3rd also with 1st, and so on...?

**Comment by bhirsz** [comment]
No, every time we would run ``compare_runs`` report we would produce (and overwrite if already exist) ``.robocop_results`` file. That's why it will produce:

- 1st run would end up with "No previous run to compare. Run robocop again to compare results" and we would produce the file
- 2nd run would compare with 1st, and overwrite the file
- 3rd run would compare with 2nd, and overwrite the file
- etc

**Comment by mnojek** [comment]
Sure, but if you do not run the `compare_runs` report for some time, then when you run it again, it will compare the results with the last created report, which may produce the difference that would not include the runs in between, right?

In such case, if we will use the `.robocop` directory that will produce the output anyway, we would be able to compare each runs between themselves and also always be able to compare last 2 runs, right?

**Comment by bhirsz** [comment]
Yes exactly, if user would run Robocop without --reports compare_runs it will not produce file, so the later difference would miss those runs.

Depends on how we implement it. We can just produce last result file (and we can always be able to compare last 2 runs), or we can create separate file for each run and then we could either compare last 2 runs or in any combination (first with last). But that would introduce some complexity, we would need extra configuration etc but is certainly doable.

**Comment by bhirsz** [comment]
As we discussed:

 - there should be ``--persist`` flag to save the reports summaries to file
 - by default we can try to use system cache:
 
```
Linux: "~/.cache/robocop"
Mac OS X: "~/Library/Caches/robocop"
Windows: "C\:Users\<username>\AppData\Local\robocop"
```

Summary is always saved to file with ``--persist`` but only displayed if user add ``compare_runs`` report. This is special report that doesn't display anything but "enhance" output of the other reports. For example:

```
Processed 314 files from which 304 (-3) files contained issues.

Found 3610(-10) issues: 2450 (-4) WARNINGs, 470 INFOs, 690 (-6) ERRORs.

Issues by ID:
W0201 (missing-doc-keyword)                        : 399 (+2)
W0202 (missing-doc-test-case)                      : 237 (-10)
...

Scan finished in 7.342s (- 1.2s)
```

**Comment by mnojek** [comment]
Additionally, if the flag `--persist` is not set, when `compare_runs` is called, it should say that there is no collected data, or it's old and that the `--persist` flag needs to be turned on as well.

Ideally, I imagine a user having a configuration in a file with `persist=yes` or something like that, and then just calling `--reports all,compare_runs` should display all the reports with a comparison with a previous run.

Later, we can try to add colors to the output, so that the `(+10)` would be red and `(-6)` green, for the number of detected issues, for example.

**Comment by bhirsz** [comment]
Example of the compare between full run, and run with all rules disabled:

```
Processed 137 (+0) files but no issues were found. Previously there were 77 files with issues.

Found 0 (-253) issues.

Report generated by Robocop version: 3.2.1

Issues by ID:

I0923 (unnecessary-string-conversion)     : 0 (-57)
I0913 (can-be-resource-file)              : 0 (-49)
W0701 (todo-in-comment)                   : 0 (-37)
W0503 (too-many-calls-in-keyword)         : 0 (-34)
W0505 (too-many-calls-in-test-case)       : 0 (-19)
W0501 (too-long-keyword)                  : 0 (-10)
W0507 (too-many-arguments)                : 0 (-8)
W0919 (unused-argument)                   : 0 (-7)
W0506 (file-too-long)                     : 0 (-7)
I0924 (expression-can-be-simplified)      : 0 (-6)
I0920 (unused-variable)                   : 0 (-6)
I0316 (possible-variable-overwriting)     : 0 (-4)
W0504 (too-long-test-case)                : 0 (-3)
W0922 (variable-overwritten-before-usage) : 0 (-2)
W0301 (not-allowed-char-in-name)          : 0 (-2)
W0324 (overwriting-reserved-variable)     : 0 (-1)
I0605 (could-be-test-tags)                : 0 (-1)

Reported: 2023-06-29 15:33:39 +0200

Scan finished in 1.349s (-2.919).
```

Another comparison on single file:
```
Processed 1 (+0) file from which 1 (+0) file contained issues.

Found 2 (+0) issues: 1 (+0) INFO, 1 (+0) WARNING.

Report generated by Robocop version: 3.2.1

Issues by ID:

I0913 (can-be-resource-file) : 1 (+0)
W0919 (unused-argument)      : 1 (+0)

Reported: 2023-06-29 15:35:17 +0200

Scan finished in 0.018s (+0.003).

```

**Comment by bhirsz** [comment]
Cached file also saves the results per work directory (so it's possible to run Robocop on different project and the results will not be lost).

**Comment by mnojek** [comment]
Amazing stuff! Looks great 😎
